include/net/can.h

```diff
@@ -0,0 +1,186 @@
+ */
+
+/*
+ * Copyright (c) 2019 Alexander Wachter
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_NET_CAN_H_
+#define ZEPHYR_INCLUDE_NET_CAN_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <zephyr/types.h>
+#include <net/net_ip.h>
+#include <net/net_if.h>
+#include <can.h>
+
+/**
+ * @brief IPv6 over CAN library
+ * @defgroup net_can Network Core Library
+ * @ingroup networking
+ * @{
+ */
+
+/**
+ * CAN L2 driver API. Used by 6loCAN.
+ */
+
+/** @cond INTERNAL_HIDDEN */
+
+#define NET_CAN_DL 8
+/*0x3DFF - bit 4 to 10 must not be zero. Also prevent stuffing bit*/
+#define NET_CAN_MULTICAST_ADDR 0x3DFF
+#define NET_CAN_DAD_ADDR       0x3DFE
+#define NET_CAN_MAX_ADDR       (0x3DF0 - 1)
+#define NET_CAN_MIN_ADDR       (0x00FF + 1)
+
+#define CAN_NET_IF_ADDR_MASK       0x3FFF
+#define CAN_NET_IF_ADDR_BYTE_LEN   2
+#define CAN_NET_IF_ADDR_DEST_POS   14
+#define CAN_NET_IF_ADDR_DEST_MASK  (CAN_NET_IF_ADDR_MASK << CAN_NET_IF_ADDR_DEST_POS)
+#define CAN_NET_IF_ADDR_SRC_POS    0
+
+#define CAN_NET_FILTER_NOT_SET -1
+
+/** @endcond */
+
+/*
+ *             +-----------+           +-----------+
+ *             |           |           |           |
+ *             |   IPv6    |           |   IPv6    |
+ *             |  6LoCAN   |           |  6LoCAN   |
+ *             |           |           |           |
+ *             +----+-+----+           +----+-+----+
+ *                  | |                     | |                   +---+
+ *  +----+          | |                     | |                  /     \   +-+
+ *  |    |          | |                     | |          +--+   /       \_/   \
+ * +++   +---+   +--+----+   +-------+   +--+----+   +--/    \_/               \
+ * | |        \ /     |   \ /         \ /     |   \ /  /                       |
+ * | |         X      |    X    CAN    X      |    X  |         Internet       |
+ * | |        / \     |   / \         / \     |   / \  \                       /
+ * +++   +---+   +----+--+   +-------+   +----+--+   +--+                     /
+ *  |    |                                               +-------------------+
+ *  +----+
+ */
+struct net_can_api {
+	/**
+	 * The net_if_api must be placed in first position in this
+	 * struct so that we are compatible with network interface API.
+	 */
+	struct net_if_api iface_api;
+
+	/** Send a single CAN frame */
+	int (*send)(struct device *dev, const struct zcan_frame *frame,
+		    can_tx_callback_t cb, void *cb_arg, s32_t timeout);
+	/** Attach a filter with it's callback */
+	int (*attach_filter)(struct device *dev, can_rx_callback_t cb,
+			     void *cb_arg, const struct zcan_filter *filter);
+	/** Detach a filter */
+	void (*detach_filter)(struct device *dev, int filter_id);
+	/** Enable or disable the reception of frames for net CAN */
+	int (*enable)(struct device *dev, bool enable);
+};
+
+/** @cond INTERNAL_HIDDEN */
+
+/**
+ * Context for canbus net device.
+ */
+struct canbus_net_ctx {
+	int dad_filter_id;
```

Comment:
Please document these. Also it might be better to put dad_work struct first and then others (could avoid holes although I am not claiming there are holes here).